blog/2024-01-24-transforming-your-data-1/index.md

```diff
@@ -0,0 +1,437 @@
+    conn.execute(
+        """
+        CREATE NODE TABLE
+            Client(
+                client_id INT64,
+                name STRING,
+                age INT64,
+                PRIMARY KEY (client_id)
+            )
+        """
+    )
+
+def create_city_node_table(conn: kuzu.Connection) -> None:
+    conn.execute(
+        """
+        CREATE NODE TABLE
+            City(
+                city_id INT64,
+                city STRING,
+                PRIMARY KEY (city_id)
+            )
+        """
+    )
+
+def create_company_node_table(conn: kuzu.Connection) -> None:
+    conn.execute(
+        """
+        CREATE NODE TABLE
+            Company(
+                company_id INT64,
+                type STRING,
+                company STRING,
+                PRIMARY KEY (company_id)
+            )
+        """
+    )
+
+def create_merchant_node_table(conn: kuzu.Connection) -> None:
+    conn.execute(
+        """
+        CREATE NODE TABLE
+            Merchant(
+                merchant_id INT64,
+                company_id INT64,
+                city_id INT64,
+                PRIMARY KEY (merchant_id)
+            )
+        """
+    )
+```
+
+Note that `PRIMARY KEY` constraints are required on every node table in Kùzu, as they are used to
+ensure that edges are always created on unique node pairs. In this case, we use the `client_id`,
+`city_id`, `company_id` and `merchant_id` columns as the primary keys for each respective table.
+
+### Edge files
+
+The graph schema we designed above requires us to add new CSV files that represent the edges in
+the graph. We'll create the following edge files:
+
+- `transacted_with.csv`: connects a client to a merchant
+- `has_instance.csv`: connects a company to a merchant
+- `located_in.csv`: connects a merchant to a city
+
+We first define the empty table schemas using the `FROM` and `TO` keywords that indicate the
+direction of the edge. The names of the tables: `TransactedWith`, `HasInstance` and `LocatedIn`
+are the relationship type.
+
+```py
+def create_edge_tables(conn: kuzu.Connection) -> None:
+    conn.execute(
+        """CREATE REL TABLE TransactedWith(
+                FROM Client TO Merchant,
+                amount_usd FLOAT,
+                timestamp TIMESTAMP
+            )
+        """)
+    conn.execute("CREATE REL TABLE LocatedIn(FROM Merchant TO City)")
+    conn.execute("CREATE REL TABLE HasInstance(FROM Company TO Merchant)")
+```
+
+In the above code, only `TransactedWith` edges have metadata associated with them, namely the
+transaction amount and the timestamp. The other two edges don't have metadata and are simply
+used to connect the nodes based on the values that match a primary key constraint.
+
+An example is shown below. Consider the node tables `client.csv` and `transaction.csv` as follows:
+
+<div class="img-center">
+<img src={EdgeTables}/>
+</div>
+
+A simple transformation function can convert the transaction table into the `TransactedWith` edge
```

Comment:
Yup, I clarified with Ziyi and I now express the data transformation and query logic in pure Cypher (no Python code). If the user wants to see the end-to-end example code, that can be referenced in the GitHub repo and run via Python.

Reply:
BTW, right now, we do have a slightly tedious workflow to reshape data from existing CSV/parquet into a node/edge format:

1. `LOAD FROM` an existing CSV/parquet file to scan data
2. `COPY TO` a new CSV/parquet file
3. `COPY FROM` the new CSV/parquet file to a Kùzu table

For a better developer experience during ETL, we could bypass step 2 in the above workflow and potentially directly pass the results of a scan (in memory) to a Kùzu table, so that we don't have to write the intermediate files for simple transformation cases like these.

I'll speak to the team and articulate this a bit more in a separate issue.